File: String/Hard/Find_Substring_With_Given_Hash_Value.py
# Leetcode: https://leetcode.com/problems/find-substring-with-given-hash-value/
"""
        (a+b) % modulo = ( (a % modulo) + (b % modulo) ) % modulo;
        (a-b) % modulo = ( (a % modulo) - (b % modulo) + modulo ) % modulo;
        (a*b) % modulo = ( (a % modulo) * (b % modulo) ) % modulo;
        (a/b) % modulo = ( (a % modulo) * ( (b^-1) % modulo) ) % modulo;
"""
class Solution:
    def subStrHash(self, s: str, power: int, modulo: int, k: int, hashValue: str) -> str:
        hash = 0
        product = 0
        powerData = 1
        addCharacter = ''
        addCharacterValue = 0
        deleteCharacter = ''
        deleteCharacterValue = 0
        i = len( s ) - 1
        result = ''
        dictionary = {}
        for item in range(97, 123):
                dictionary[chr(item)] = item - 96
                
        for _ in reversed(s):
                if( i < k - 1 ):
                        break
                substring = s[i - k + 1 : i + 1]
                if( i != len( s ) - 1 ):
                        addCharacter = substring[0]
                        deleteCharacter = s[ i + 1 ]
                if( i == len( s ) - 1 ):
                        for j in range(0, len(substring)):
                                product = ( ( dictionary[ substring[j] ] % modulo ) * ( powerData % modulo ) ) % modulo
                                hash = ( ( hash % modulo ) + ( product % modulo ) ) % modulo
                                powerData = ( ( powerData % modulo ) * ( power % modulo ) ) % modulo
                else:
                        deleteCharacterValue = dictionary[ deleteCharacter ]
                        deleteCharacterValue = ( ( deleteCharacterValue % modulo ) * ( powerData % modulo ) ) % modulo
                        
                        # Multipy hash by power
                        hash = ( ( hash % modulo ) * ( power % modulo ) ) % modulo
                        
                        addCharacterValue = dictionary[ addCharacter ]
                        # Add addCharacterValue to hash
                        hash = ( ( hash % modulo ) + ( addCharacterValue % modulo ) ) % modulo
                        
                        hash = ( ( hash % modulo ) - deleteCharacterValue + modulo ) % modulo
                if( hash == hashValue ):
                        result = substring
                i -= 1
        
        return result

File: String/Hard/Find_Substring_With_Given_Hash_Value.js
// Leetcode: https://leetcode.com/problems/find-substring-with-given-hash-value/

/**
 * @param {string} s
 * @param {number} power
 * @param {number} modulo
 * @param {number} k
 * @param {number} hashValue
 * @return {string}
 */

const subStrHash = (s, p, m, k, hashValue) => {
    p = BigInt(p);
    m = BigInt(m);
    hashValue = BigInt(hashValue);
    let currentHash = BigInt(0);
    let result = 0;
    let pk = BigInt(1);
    let n = s.length;

    for (let i = n - 1; i >= 0; i--) {
        currentHash = (currentHash * p + BigInt(s[i].charCodeAt() - 96)) % m;
        if (i + k >= n) {
            pk = (pk * p) % m;
        } else {
            currentHash = (currentHash - ((BigInt(s[i + k].charCodeAt() - 96) * pk) % m) + m) % m;
        }
        if (currentHash === hashValue) {
            result = i;
        }
    }

    return s.slice(result, result + k);
};
